packages/hash/frontend/src/lib/entities.ts

```diff
@@ -1,127 +1,460 @@
+        undefined,
+        2,
+      )}`,
+    );
+  }
+
+  return identifierObject;
+}
+
+type ApiLinkIdentifier = {
+  accountId: string;
+  linkId: string;
+};
+
+/**
+ * Create a linkId that's a stringified object containing the fields we will need later.
+ * The re-written linkId is what should be sent to the block with any Link
+ */
+const rewriteLinkIdentifier = ({ accountId, linkId }: ApiLinkIdentifier) =>
+  JSON.stringify({ accountId, linkId });
+
+/**
+ * We send blocks a 'linkId' that is a stringified object in {@link convertApiLinkToBpLink}
+ * – this reverses the process so we have the individual fields to use in calling the HASH API.
+ *
+ * @param stringifiedLinkId a linkId' or equivalent (e.g. sourceEntityId) sent from a block
+ */
+export const parseLinkIdentifier = (
+  stringifiedLinkId: string,
+): ApiLinkIdentifier => {
+  let identifierObject: unknown;
+  try {
+    identifierObject = JSON.parse(stringifiedLinkId);
+  } catch (err) {
+    throw new Error(
+      `Provided identifier string '${stringifiedLinkId}' cannot be parsed to JSON: ${err}`,
+    );
+  }
+
+  if (!isObject(identifierObject)) {
+    throw new Error(
+      `Parsed link identifier is not an object: ${JSON.stringify(
+        identifierObject,
+      )}`,
+    );
+  }
+
+  if (!hasKeyForString("accountId", identifierObject)) {
+    throw new Error(
+      `Parsed identifier for Link does not contain accountId key. Provided identifier: ${JSON.stringify(
+        identifierObject,
+        undefined,
+        2,
+      )}`,
+    );
+  }
+
+  if (!hasKeyForString("linkId", identifierObject)) {
+    throw new Error(
+      `Parsed identifier for Link does not contain linkId key. Provided identifier: ${JSON.stringify(
+        identifierObject,
+        undefined,
+        2,
+      )}`,
+    );
+  }
+
+  return identifierObject;
+};
+
+/**
+ * Converts a Link from its GraphQL API representation to its Block Protocol representation,
+ * by re-writing all of linkId, sourceEntityId and sourceDestinationId so that they are
+ * stringified objects containing the identifiers we need (i.e. to include accountId)
+ */
+export const convertApiLinkToBpLink = ({
+  linkId,
+  destinationAccountId,
+  destinationEntityId,
+  sourceAccountId,
+  sourceEntityId,
+  path,
+  index,
+}: Pick<
+  ApiLink,
+  | "linkId"
+  | "destinationAccountId"
+  | "destinationEntityId"
+  | "sourceAccountId"
+  | "sourceEntityId"
+  | "index"
+  | "path"
+>): BpLink => {
+  if (linkId.includes("{")) {
+    throw new Error(
+      `linkId has already been re-written as a stringified object: ${linkId}`,
+    );
+  }
+
+  if (sourceEntityId.includes("{")) {
+    throw new Error(
+      `sourceEntityId has already been re-written as a stringified object: ${sourceEntityId}`,
+    );
+  }
+
+  if (destinationAccountId.includes("{")) {
+    throw new Error(
+      `destinationAccountId has already been re-written as a stringified object: ${destinationAccountId}`,
+    );
+  }
```

Comment:
**Blocking:** I could be misunderstanding, but shouldn't we be checking whether `destinationEntityId` is already stringified here? As far as I'm aware we're not stringifying objects into `destinationAccountId` anywhere